TACC/Core-Portal
# task/TV3-44 Notifications

## Overview

The goal for this task was to refactor the notification functionality for job status updates using the new tapis v3 responses. All of the functionality was kept the same and some fields/code that were not being used were cleaned up and removed.

## Related

* [TV3-44](https://jira.tacc.utexas.edu/browse/TV3-44)
* [TV3-54](https://jira.tacc.utexas.edu/browse/TV3-54)
* [TV3-75](https://jira.tacc.utexas.edu/browse/TV3-75)

## Changes

- Refactored code to use the updated fields from the [tapis v3 response](https://tapis.readthedocs.io/en/latest/technical/jobs.html#notification-messages)
- Changed valid job statuses based on [v3 docs](https://tapis.readthedocs.io/en/latest/technical/jobs.html#job-status)
- Notification object no longer includes the entire job object and instead only includes limited required information for a job
- Removed some unused fields from the notification object that were not being used in the portal. Fields removed include: 
    1. action_link: This field was removed from the tapis job notification object but is still kept for interactive job notifications.
    2. target_path: This field was not used anywhere in the portal
- No changes made to interactive webhook handler except for a minor function name change

## Testing

1. Log on to the portal 
3. Create a test job (I have been using the hello-world application from [WMA Tapis Templates](https://github.com/TACC/WMA-Tapis-Templates) via tapipy to create jobs)
4. Ensure job notifications are received in the portal and status is updated correctly. 
5. Ensure job is correctly shown in the History tab
6. Ensure Output Location column updates correctly after a job reaches a terminal state (Finished, Cancelled, Failed)

## UI

- No UI changes made

## Notes

A couple of things to note that I am not entirely sure about: 

1. For the job output location I kept functionality the same where it is just the value of `archiveSystemId/archiveSystemDir`. Based on [this slack conversation](https://tacc-team.slack.com/archives/C0307MGPCP2/p1668810929775119?thread_ts=1668790809.825759&cid=C0307MGPCP2) with Richard he mentioned that this might not be a reliable way to get the job output location as it can be pretty dynamic and instead recommended to use the [getJobOutputList](https://tapis.readthedocs.io/en/latest/technical/jobs.html#get-job-output-listing) query to get this path. Wanted to get the teams thoughts on this. 

2. I haven't been able to properly test the [job indexing part](https://github.com/TACC/Core-Portal/blob/task/TV3-44-notifications/server/portal/apps/webhooks/views.py#L173) after a job finishes since I have been running into some job errors but I've kept functionality the same so testing this and making sure it works would be important.
